crates/bevy_ecs/src/schedule/graph/tarjan_scc.rs

```diff
@@ -0,0 +1,222 @@
+use super::DiGraph;
+use super::NodeId;
+use alloc::vec::Vec;
+use core::hash::BuildHasher;
+use core::num::NonZeroUsize;
+use smallvec::SmallVec;
+
+/// Create an iterator over *strongly connected components* using Algorithm 3 in
+/// [A Space-Efficient Algorithm for Finding Strongly Connected Components][1] by David J. Pierce,
+/// which is a memory-efficient variation of [Tarjan's algorithm][2].
+///
+///
+/// [1]: https://homepages.ecs.vuw.ac.nz/~djp/files/P05.pdf
+/// [2]: https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm
+///
+/// Returns each strongly strongly connected component (scc).
+/// The order of node ids within each scc is arbitrary, but the order of
+/// the sccs is their postorder (reverse topological sort).
+pub(crate) fn new_tarjan_scc<S: BuildHasher>(
+    graph: &DiGraph<S>,
+) -> impl Iterator<Item = SmallVec<[NodeId; 4]>> + '_ {
+    // Create a list of all nodes we need to visit.
+    let unchecked_nodes = graph.nodes();
+
+    // For each node we need to visit, we also need to visit its neighbours.
+    // Storing the iterator for each set of neighbours allows this list to be computed without
+    // an additional allocation.
+    let nodes = graph
+        .nodes()
+        .map(|node| NodeData {
+            rootindex: None,
+            neighbours: graph.neighbors(node),
+        })
+        .collect::<Vec<_>>();
+
+    TarjanScc {
+        graph,
+        unchecked_nodes,
+        index: 1,                   // Invariant: index < componentcount at all times.
+        componentcount: usize::MAX, // Will hold if componentcount is initialized to number of nodes - 1 or higher.
+        nodes,
+        stack: Vec::new(),
+        visitation_stack: Vec::new(),
+        start: None,
+        indexadjustment: None,
+    }
+}
+
+struct NodeData<N: Iterator<Item = NodeId>> {
+    rootindex: Option<NonZeroUsize>,
+    neighbours: N,
+}
+
+/// A state for computing the *strongly connected components* using [Tarjan's algorithm][1].
+///
+/// This is based on [`TarjanScc`] from [`petgraph`].
+///
+/// [1]: https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm
+/// [`petgraph`]: https://docs.rs/petgraph/0.6.5/petgraph/
+/// [`TarjanScc`]: https://docs.rs/petgraph/0.6.5/petgraph/algo/struct.TarjanScc.html
+struct TarjanScc<'graph, Hasher, AllNodes, Neighbours>
+where
+    Hasher: BuildHasher,
+    AllNodes: Iterator<Item = NodeId>,
+    Neighbours: Iterator<Item = NodeId>,
+{
+    /// Source of truth [`DiGraph`]
+    graph: &'graph DiGraph<Hasher>,
+    /// An [`Iterator`] of [`NodeId`]s from the `graph` which may not have been visited yet.
+    unchecked_nodes: AllNodes,
+    /// The index of the next SCC
+    index: usize,
+    /// A count of potentially remaining SCCs
+    componentcount: usize,
```

Comment:
why not `component_count`

Reply:
No particular reason, just a hangover from the original `petgraph` reference implementation. I agree with these name changes!